sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -205,10 +205,14 @@ object ConstantPropagation extends Rule[LogicalPlan] {
     def replaceConstants0(expression: Expression) = expression transform {
       case a: AttributeReference => constantsMap.getOrElse(a, a)
     }
-    condition transform {
-      case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants0(e)
-      case e @ EqualNullSafe(_, _) if !predicates.contains(e) => replaceConstants0(e)
-    }
+    splitConjunctivePredicates(condition).map {
+      // Don't replace constants in IsNotNull because InferFiltersFromConstraints may infer
+      // another IsNotNull
+      case isNotNull: IsNotNull => isNotNull
+      case e: EqualTo if predicates.contains(e) => e
+      case e: EqualNullSafe if predicates.contains(e) => e
```

Comment:
previously we can optimize `EqualTo(...) Or EqualTo(...)` because we use `condition.transform`. Seems it's not the case now after your change?

Reply:
It dose not support this case: `a = b + 1 or b = 10`. 
It supports this case: `(a = b + 1 and b = 9) or b = 10` => `(a = 10 and  b = 9) or b = 10)`.

This change will not break the current behavior.

Reply:
It is incorrect if still use `transform` to support other predicates:
```diff
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
index 4542ecb95f..90f8c5849b 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
@@ -109,7 +109,7 @@ object ConstantFolding extends Rule[LogicalPlan] {
  * - Using this mapping, replace occurrence of the attributes with the corresponding constant values
  *   in the AND node.
  */
-object ConstantPropagation extends Rule[LogicalPlan] {
+object ConstantPropagation extends Rule[LogicalPlan] with PredicateHelper {
   def apply(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
     _.containsAllPatterns(LITERAL, FILTER), ruleId) {
     case f: Filter =>
@@ -208,6 +208,7 @@ object ConstantPropagation extends Rule[LogicalPlan] {
     condition transform {
       case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants0(e)
       case e @ EqualNullSafe(_, _) if !predicates.contains(e) => replaceConstants0(e)
+      case other => replaceConstants0(other)
     }
   }
 }
```

For example:
```
Step 1: `a = b + 1 and b = 10` ------> It will go to other branch
Step 2: `a = 10 + 1 and 10 = 10`
Step 3: `a = 11 and true`
Step 4: `a = 11`
```

Reply:
I'm not convinced that the new code can cover all the cases of the old code. e.g. `Not(EqualTo)` can be handled by `condition.transform`

Reply:
It will not will not break the current behavior. I have added 2 `Not` related tests:
https://github.com/apache/spark/blob/58d8f20558a7365edcbd32717dc6a76bf703bf5e/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantPropagationSuite.scala#L99-L115

Reply:
I'm not asking for tests, I'm asking for an explanation...

Reply:
It still works for the following expressions:
```
not(a = d + 1 && d = 9)
not(a = b + 1) && b = 10
```